vmdb/app/models/foreman_provision_task.rb

```diff
@@ -0,0 +1,66 @@
+require 'manageiq_foreman'
+class ForemanProvisionTask < MiqRequestTask
+  # provisioning builds the box
+  # vs alternative that just sets the attributes and doesn't rebuild
+  def build?
+    options.key?(:build) ? options[:build] : true
+  end
+
+  def power_off
+    source.with_provider_object(&:stop)
+    signal :powering_off
+  end
+
+  def powering_off
+    if !source.with_provider_object(&:powered_on?)
+      signal :reboot_source
+    else
+      requeue_phase
+    end
+  end
+
+  def reboot_source
+    source.with_provider_object(&:reboot)
+    signal :power_on
+  end
+
+  def power_on
+    source.with_provider_object(&:start)
+    signal :building
+  end
+
+  def building
+    if confirm_not_building?
+      signal :final_refresh
+    else
+      requeue_phase
+    end
+  end
+
+  def confirm_not_building?
+    refresh # actively queue a refresh, ok?
+    !source.pending?
+  end
+
+  def refresh
+    EmsRefresh.queue_refresh(source)
+  end
+
+  def final_refresh
+    refresh # actively queue a refresh, ok?
+    signal :provision_complete
+  end
+
+  def provision_complete
+  end
+
+  def run_provision
+    puts "Updating build flag to #{build?} and hostgroup_id for #{source.hostname}"
+    source.with_provider_object { |p| p.update(options) }
```

Comment:
We have logic in other state machines like ```log_clone_options``` that I would also expect to have here.  It allows the users to see the changes we plan to make before we send them across to the provider.  The field finds this very helpful in troubleshooting failures.